File: src/main/storage/SettingsStore.ts
import fs from 'node:fs'
import path from 'node:path'
import { app } from 'electron'
import {
  SettingsStateSchema,
  settingsVersion,
  type CleanupHistoryEntry,
  type InstallHistoryEntry,
  type SettingsState
} from '@shared/contracts'

const defaultSettings: SettingsState = {
  version: settingsVersion,
  wizardCompleted: false,
  recentCleanupActions: [],
  recentInstalls: []
}

export class SettingsStore {
  private readonly filePath: string
  private cache: SettingsState

  constructor() {
    this.filePath = path.join(app.getPath('userData'), 'settings.json')
    this.cache = this.load()
  }

  get(): SettingsState {
    return this.cache
  }

  setWizardCompleted(value: boolean): SettingsState {
    return this.update({
      wizardCompleted: value
    })
  }

  setAdbPath(adbPath?: string): SettingsState {
    return this.update({
      adbPath,
      wizardCompleted: false
    })
  }

  setManualEndpoint(manualEndpoint?: string): SettingsState {
    return this.update({
      manualEndpoint,
      wizardCompleted: false
    })
  }

  recordInstall(entry: InstallHistoryEntry): SettingsState {
    const recentInstalls = [entry, ...this.cache.recentInstalls.filter((item) => item.packageName !== entry.packageName)].slice(0, 50)
    return this.update({
      recentInstalls
    })
  }

  recordCleanup(entry: CleanupHistoryEntry): SettingsState {
    const recentCleanupActions = [entry, ...this.cache.recentCleanupActions].slice(0, 50)
    return this.update({
      recentCleanupActions
    })
  }

  private update(patch: Partial<SettingsState>): SettingsState {
    this.cache = SettingsStateSchema.parse({
      ...this.cache,
      ...patch
    })

    fs.mkdirSync(path.dirname(this.filePath), { recursive: true })
    fs.writeFileSync(this.filePath, JSON.stringify(this.cache, null, 2), 'utf8')
    return this.cache
  }

  private load(): SettingsState {
    try {
      if (!fs.existsSync(this.filePath)) {
        return defaultSettings
      }

      const raw = fs.readFileSync(this.filePath, 'utf8')
      const parsed = JSON.parse(raw)
      return SettingsStateSchema.parse({
        ...defaultSettings,
        ...parsed
      })
    } catch {
      return defaultSettings
    }
  }
}

File: WsaManager.Core/SettingsStore.cs
using System.Text.Json;

namespace WsaManager.Core;

public sealed class SettingsStore
{
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };
    private readonly string filePath;
    private SettingsState cache;

    public SettingsStore(string appDataDirectory)
    {
        filePath = Path.Combine(appDataDirectory, "settings.json");
        cache = Load();
    }

    public string AppDataDirectory => Path.GetDirectoryName(filePath)!;

    public SettingsState Get() => Clone(cache);

    public SettingsState SetAdbPath(string? adbPath)
    {
        cache.AdbPath = string.IsNullOrWhiteSpace(adbPath) ? null : adbPath;
        cache.WizardCompleted = false;
        Save();
        return Get();
    }

    public SettingsState SetManualEndpoint(string? manualEndpoint)
    {
        cache.ManualEndpoint = string.IsNullOrWhiteSpace(manualEndpoint) ? null : manualEndpoint;
        cache.WizardCompleted = false;
        Save();
        return Get();
    }

    public SettingsState SetWizardCompleted(bool value)
    {
        cache.WizardCompleted = value;
        Save();
        return Get();
    }

    public SettingsState RecordInstall(InstallHistoryEntry entry)
    {
        cache.RecentInstalls = new[] { entry }
            .Concat(cache.RecentInstalls.Where(item => item.PackageName != entry.PackageName))
            .Take(50)
            .ToList();
        Save();
        return Get();
    }

    private SettingsState Load()
    {
        try
        {
            if (!File.Exists(filePath))
            {
                return new SettingsState();
            }

        var loaded = JsonSerializer.Deserialize<SettingsState>(File.ReadAllText(filePath), JsonOptions) ?? new SettingsState();
        loaded.Version = 1;
        loaded.RecentInstalls ??= [];
        loaded.RecentInstalls.ForEach(entry =>
        {
            if (entry.SizeBytes < 0)
            {
                entry.SizeBytes = 0;
            }
        });
        return loaded;
        }
        catch
        {
            return new SettingsState();
        }
    }

    private void Save()
    {
        Directory.CreateDirectory(Path.GetDirectoryName(filePath)!);
        File.WriteAllText(filePath, JsonSerializer.Serialize(cache, JsonOptions));
    }

    private static SettingsState Clone(SettingsState settings)
    {
        return JsonSerializer.Deserialize<SettingsState>(JsonSerializer.Serialize(settings, JsonOptions), JsonOptions) ?? new SettingsState();
    }
}
